File: genwavfm.jl
#!/usr/bin/env julia

# Parse arguments
inputpath = ARGS[2]
outputpath = ARGS[4]

# Ensure directory paths end in slashes
if inputpath[end] != '/'
    inputpath = inputpath * "/"
end

if outputpath[end] != '/'
    outputpath = outputpath * "/"
end

# Main part of script
for (root, dirs, files) in walkdir(inputpath)
    mp3s = filter(x -> contains(x, "mp3"), files)
    for mp3 in mp3s
        # Set artist and album variables; ffpath for system commands
        artist = basename(dirname(root))
        album = basename(root)
        ffpath = joinpath(root, mp3)

        # Get length of mp3 via ffprobe
        length = readstring(`ffprobe -v error -show_entries format=duration $ffpath`)
        length = match(r"\d+\.\d+", length)
        length = ceil(float(length.match))

        if artist == "Various Artists" # Soundtracks/compilations
            output_filename = "VA_" * album * "_" * mp3
        else # Everything else
            output_filename = artist * "_" * album * "_" * mp3
        end
        output_filename = replace(output_filename, " ", "_")
        output_filename = replace(output_filename, "mp3", "png")
        outputP = joinpath(outputpath, output_filename)

        run(`audiowaveform -i $ffpath -o $outputP -e $length -w 1600 -h 500`)
    end
end

File: genwavfm.lua
require 'lfs'

-- Parse arguments
inputpath = arg[2]
outputpath = arg[4]

-- Ensure directory paths end in slashes
if string.sub(inputpath, -1) ~= '/' then
   inputpath = inputpath .. '/'
end

if string.sub(outputpath, -1) ~= '/' then
   outputpath = outputpath .. '/'
end

-- Main part of script
for artist in lfs.dir(inputpath) do
   if artist ~= "." and artist ~= ".." then
      for album in lfs.dir(inputpath .. artist) do
         if album ~= "." and album ~= ".." then
            for mp3 in lfs.dir(inputpath .. artist .. '/' .. album) do
               if mp3:match("%.mp3$") then
                  ffpath = inputpath .. artist .. '/' .. album .. '/' .. mp3
                  ch = io.popen('ffprobe -v error -show_entries format=duration "' .. ffpath .. '"')
                  result = ch:read('*a')
                  length = string.match(result, "%d+.%d+")
                  length = math.ceil(length)

                  if artist == 'Various Artists' then
                     outputF = 'VA_' .. album .. '_' .. mp3
                  else
                     outputF = artist .. '_' .. album .. '_' .. mp3
                  end
                  outputF = string.gsub(outputF, ' ', '_')
                  outputF = string.gsub(outputF, 'mp3', 'png')
                  outputF = outputpath .. outputF

                  os.execute('audiowaveform -i "' .. ffpath .. '" -o ' .. outputF .. ' -e ' .. length .. ' -w 1600 -h 500')
               end
            end
         end
      end
   end
end
